### ISSUE TITLE
[Android] CollectionViewwith multiple bindable layouts (CardWithComplexContent) performance issues

### ISSUE THREAD

**Issue by eth-ellis** [dotnet/maui#21580]
### Description

When migrating our app from Xamarin.Forms to .NET MAUI we noticed that for most of our CollectionViews the performance when scrolling was worse.

When new views appeared the CollectionView would stutter/jitter resulting in an unpleasant user experience.

Our item templates are somewhat complex but still worked great in Xamarin.Forms.

In the linked repro apps, we have a template called _CardWithComplexContent_ which is somewhat similar to the template in our app.

**Note**
- In the .NET MAUI project we are using nightly build 8.0.20-nightly.10367
  - This includes https://github.com/dotnet/maui/pull/21229
- We are using **Release** configuration when testing performance

### Android (Samsung A73) - Xamarin.Forms

https://github.com/dotnet/maui/assets/13865151/25eac924-a499-4908-afd9-ab7108d71437

### Android (Samsung A73) - MAUI

https://github.com/dotnet/maui/assets/13865151/cd402208-674f-4cd4-8a67-b3038e1e8f88

### iOS (iPhone 6s) - Xamarin.Forms

https://github.com/dotnet/maui/assets/13865151/13a64912-3daf-4325-9d0e-4c15d4cee2eb

### iOS (iPhone 6s) - MAUI

https://github.com/dotnet/maui/assets/13865151/d46972ec-5e51-479d-aa53-81351874db9e

### Steps to Reproduce

1. Checkout .NET MAUI and Xamarin.Forms repro projects.
2. Build and deploy the apps to your devices in **Release** configuration.
3. Run both apps and select either the _CardWithTheLot_ or _CardWithComplexContent_ template.
4. Compare scrolling performance on both apps (fast scrolling and slow scrolling).

The apps contain other templates for attempting to isolate elements which impact performance the most.

Accepting PRs or suggestions on the repro app repos for templates to add for comparison.

### Acceptance Criteria
The scrolling experience on .NET MAUI performs as well as or better than Xamarin.Forms for all affected platforms.

### Link to public reproduction project repository

https://github.com/eth-ellis/Issue-Repro/tree/main/CollectionViewPerformanceXamarin and https://github.com/eth-ellis/Issue-Repro/tree/main/CollectionViewPerformanceMaui

### Version with bug

Nightly / CI build (Please specify exact version)

### Is this a regression from previous behavior?

Yes, this used to work in Xamarin.Forms

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, Android

**Comment by Matt-17** [comment]
Unfortunately I must agree. I just put 100 same size labels in one collectionView and it's stuttering on my Pixel 5. In Forms this was totally fine. My workaround is now putting this in VerticalStackLayout and ScrollView as I don't need the other extras. I loads slower, but better than this scrolling experience. 
Bonus points: put this CollectionView inside a CarouselView and you're crying.

**Comment by jonmdev** [comment]
Thank you for making this thread and providing such objective evidence. Your post clearly shows this is a Maui-specific problem. 

For added information, I built my own version of a "CollectionView" which performs similarly poorly, and have also created simple simulations of simple translation/update of elements to try to figure out the issue, all of which behave the same. 

I have tried two methods of creating/simulating my own "CollectionView" type function:

1) Having 10-15 elements that each translate directly themselves together on screen and then wrap top to bottom to simulate scrolling. I posted a simple simulation of this nature [here](https://github.com/dotnet/maui/issues/20979). The elements are derived from AbsoluteLayout and hold each only Border/Image/Label elements.

2) Having 10-15 similar elements on an AbsoluteLayout as a base which translates (so I am not translating the elements themselves each frame but rather their parent AbsoluteLayout) and then wrapping/updating the necessary elements only as they hit screen top or bottom.

In both cases the performance on Android has been identical to what you show in your videos. Even if I simplify the display elements so they are just a few labels on an AbsoluteLayout, Android struggles to update them effectively in Maui. I can't build release iOS versions to compare.  

(I closed my linked report as I thought perhaps I needed to manually freeze elements on translation in Android as mentioned in that thread to fix it. However, afterwards, I can say that didn't help actually.)

Given I can reproduce the same stuttery "scrolling" even with simple creations made of nothing more than AbsoluteLayout, Image, Label, and Border with very simple updates, I suspect the problem is more generalized in Maui than the CollectionView. 

I suspect it has something to do with Maui perhaps invalidating the platform views in a way that is making them redraw the screen too many times when even simple updates occur. There is clearly way too much effort going on. Translations should be extremely cheap to adjust and Label text updates should also be very simple. Yet they are not. Simple things like this cause the apps to grind down. Watching the Android Studio profiler when I load an APK, I can see it is doing too much work each update but don't know why.

Hopefully your evidence will help lead to the cause and solution.

**Comment by Zhanglirong-Winnie** [comment]
Verified this issue with Visual Studio 17.10.0 Preview 2(8.0.14 & 8.0.0-rc.2.9530). Can repro on android and iOS platforms with sample project.

**Comment by bcaceiro** [comment]
I am converting all my collectionviews for BindableLayouts ( since I don't have that many elements) and the performance gains are  huge.

On other note, one thing I noticed: Without interacting with the collectionview, the Garbage Collector is crying, spamming multiple  GCs - which is kinda strange? (after loading the CV)

**Comment by Alex-Dobrynin** [comment]
this is all about new layout system in MAUI. xamarin forms used another layout system. if look into the code of collectionview in maui and in xf, they are pretty the same. but the layout system had significant changes not in a good way

**Comment by MichaelShapiro** [comment]
@Alex-Dobrynin I'd be interested to look at the code. Thank you.

**Comment by jonmdev** [comment]
> this is all about new layout system in MAUI. xamarin forms used another layout system. if look into the code of collectionview in maui and in xf, they are pretty the same. but the layout system had significant changes not in a good way.
> 
> By the way, just created own virtualize listview based on scrollview and absolute layout. If somebody wants, can share the code. it does not use any row of native code.
> 
> this is android, on ios almost the same. also works on other platfroms
> video5195289640646495984.mp4

I am confused by your post. I also agree with you this is not a "CollectionView" issue but rather a Maui issue. As I noted, I created my own "CollectionView" type systems, but they all performed equally poorly.

The problem cannot be directly in the "CollectionView" as you note, but rather inefficiencies in how Maui is updating the elements/layouts or invalidating the views and forcing excess redraws. Even if the only elements you use are AbsoluteLayout elements (containing a few Borders/Labels/Images) translating on an AbsoluteLayout base (simplest Layout system possible) the problem persists.

For example, @jonathanpeppers found one problem here and fixed it with Labels: https://github.com/dotnet/maui/pull/21291

But I presume there are other major problems still. Do you know of any specific problem with the Layout that might be causing all of this?

Also are you saying you found a way to circumvent this? If the problem is with the core elements/layout system then I am not sure how it can be circumvented without skipping Maui altogether.

**Comment by brentpbc** [comment]
@Alex-Dobrynin I would also be very interested to view the code.

**Comment by FM1973** [comment]
@Alex-Dobrynin me too

**Comment by WebGoose** [comment]
@Alex-Dobrynin me too

**Comment by IrynaDoroshenkoDev** [comment]
@Alex-Dobrynin looks pretty good and raelly fast!

**Comment by Alex-Dobrynin** [comment]
this is not finished version, but it supports different scroll orientations and device orientation change. still working on it
[VirtualizeListView.zip](https://github.com/dotnet/maui/files/14890022/VirtualizeListView.zip)

**Comment by mjsb212** [comment]
> this is all about new layout system in MAUI. xamarin forms used another layout system. if look into the code of collectionview in maui and in xf, they are pretty the same. but the layout system had significant changes not in a good way.
> 

I absolutely concur with the others here who are suggesting this has something to do with MAUI specifically and the Layout System. I have built my own CollectionView, and have also used 2 Third party versions (Telerik, Syncfusion) and I get the same exact results in every scenario.  After doing all of that, and testing and profiling, and stepping through source... I can see there are issues with sizing and recycling when using certain layouts and templates -- most notably when using images and grid layouts in CollectionView, the performance is terrible when scrolling no mater what vendor or version of CollectionView used. There have been a few existing open and backlogged issues about this for a while, and everything seems to be related, so I'm glad this is finally being looked in to.

**Comment by mjsb212** [comment]
> this is all about new layout system in MAUI. xamarin forms used another layout system. if look into the code of collectionview in maui and in xf, they are pretty the same. but the layout system had significant changes not in a good way.

I absolutely concur with the others here who are suggesting this has something to do with MAUI specifically and the Layout System. I have built my own CollectionView, and have also used 2 Third party versions (Telerik, Syncfusion) and I get the same exact results in every scenario. After doing all of that, and testing and profiling, and stepping through source... I can see there are issues with sizing and recycling when using certain layouts and templates -- most notably when using images and grid layouts in CollectionView; the performance is terrible when scrolling, no mater what vendor or version of CollectionView used. There have been a few existing open and backlogged issues about this for a while, and everything seems to be related, so I'm glad this is finally being looked in to.

**Comment by AliKarimiENT** [comment]
Hi,

I've been grappling with an issue that is significantly impacting the performance of our app, much to the frustration of the business owner. The problem lies with our Maui implementation, where we have a grouped list with headers and items, and a pagination feature to load additional data.

This performance issue is surprising to me, especially since I have prior experience with Flutter where I did not encounter such problems. In Flutter, we typically use a scroll view containing a ListView or Column, which automatically triggers data loading when the user scrolls to the bottom. However, attempting a similar setup in Maui has resulted in poor performance.

**Comment by Alex-Dobrynin** [comment]
@AliKarimiENT in this version no, this is just POC of smooth scrolling, but on my pc already added cells appering/dissapering functionality and load more. Need to edit layout manager to support this

**Comment by AliKarimiENT** [comment]
> @AliKarimiENT in this version no, this is just POC of smooth scrolling, but on my pc already added cells appering/dissapering functionality and load more. Need to edit layout manager to support this

I reviewed the code but I think it will take time to add paging my self, and I will wait to maui update it.

**Comment by Alex-Dobrynin** [comment]
@AliKarimiENT maui will not update it, because this is my implementation. Maui can update only their collectionview

**Comment by samhouts** [comment]
Thank you for the detailed report and reproduction projects! We're investigating this particular issue, and we'll respond with any findings. CollectionView performance is one of our top priorities, so we are working on a variety of fixes that should positively impact most cases.

**Comment by Redth** [comment]
There's a number of factors that may compound here causing this particular sample to be somewhat slower.  We are continuing to investigate these and look at introducing fixes and improvements.  Unfortunately this isn't a simple one change fixes everything.

What I can suggest in the meantime is making some changes to your XAML and data model to minimize scenarios that are going to impact performance.  For instance, I made some changes to the original sample here: https://github.com/eth-ellis/Issue-Repro/compare/main...Redth:Issue-Repro:main

Generally having a bindable layout inside of a data template is not going to be a fast thing to do.  I understand that it worked reasonably well in Xamarin.Forms, and we will be able to get there with MAUI too but for now, a bit of compromise and optimization of your layouts would help unblock this scenario.

**Comment by Alex-Dobrynin** [comment]
@Redth this is just temporary workaround. It takes a lot of time to do such changings in the templates.
But here the updated version of the VirtualListView I provided above, where it is much much performant then regular ListView and even CollectionView. And I'm sure it can be optimized much more, but I don't have enough time to play with it, for me it is enough of scrolling performance. And as you can see, there is no any platformspecific code for laying out the items. And because of that it works on all platforms out of the box without specific renderers (handlers) (except android for overscroll).

[VirtualizeListView.zip](https://github.com/dotnet/maui/files/15016088/VirtualizeListView.zip)

Also, you can write custom data adapter for grouped lists, and it is easy to add there sticky headers. Also you can add there grid items layout manager, for now only linear.

Just give it a try and play with it. may be you can take this idea to write your implementations for new CollectionView or optimize this one. But existing CollectionView is overengineering and should be rewritten completely.

**Comment by jonmdev** [comment]
> There's a number of factors that may compound here causing this particular sample to be somewhat slower. We are continuing to investigate these and look at introducing fixes and improvements. Unfortunately this isn't a simple one change fixes everything.
> 
> What I can suggest in the meantime is making some changes to your XAML and data model to minimize scenarios that are going to impact performance. For instance, I made some changes to the original sample here: [eth-ellis/Issue-Repro@main...Redth:Issue-Repro:main](https://github.com/eth-ellis/Issue-Repro/compare/main...Redth:Issue-Repro:main)
> 
> Generally having a bindable layout inside of a data template is not going to be a fast thing to do. I understand that it worked reasonably well in Xamarin.Forms, and we will be able to get there with MAUI too but for now, a bit of compromise and optimization of your layouts would help unblock this scenario.

I have not had a "bindable layout inside a data template." I am just doing straight C# and my basic label and translation updates are very slow causing similar lag and delays I don't get with similar builds in .NET Andorid. I think this is a very generalized and broad problem that runs through Maui. As you said, there are likely numerous issues.

I hope you can work on general fixes that will help us all. We all need the fast smooth scrolling users expect and every other UI system provides.

**Comment by jonmdev** [comment]
> #### I can confirm
> And this is very fast because I recycle the RowViewModels, I simply change their internal Monkey objects! Under the hood I prevent the collectionView of pushing new/recycled views on the screen.
> ### So the main problem is not the Bindings!
> ## Now it could arise from two main reasons:
> ### 1. How CollectionView assign/managed sourced item with new/recycled View.
> ### 2. How .NET MAUI render Views on the screen, even with already constructed Views.
> 
> The example provided by @Alex-Dobrynin suggests the first reason.
> 
> But in another scenario, i suggest the second.
> 
> > When I want to push a complex View (kind of a popup) in a Grid for example
> > and I use a button to Add/Remove this popup View from the Grid.
> > I can remove the View pretty fast.
> > But adding the View to the Grid depends if your finger is align with the stars ✨.
> > It can be very fast (less than 300ms) or pretty slow (more than 600ms)
> 
> ### Both reasons seems related
> 
> @Redth @samhouts @PureWeen and the all team Thank you for your time ❤️

There may be multiple reasons all at once, so maybe all these things are true. But I can absolutely agree and verify "bindings" are not the primary issue. I don't use bindings at all and still see terrible scroll performance in my custom C# "CollectionView" type design (which works blazingly fast and smooth in other non-Maui systems). 

I manage my own reassignments of elements. From a C# perspective this is basically instant as this is very simple code in my system. So again, perhaps that is poorly done in Maui CollectionView, but it is not also the main issue.

> ### 2. How .NET MAUI render Views on the screen, even with already constructed Views.

This is the primary issue without a doubt. Although I don't want to share my real proprietary code and systems (and they are baked in deep into my projects anyway to excise), I have already posted a simple "autoscroller" that this is the case and can do more if needed.

The C# code that we have control translation updates can take less than 1 ms, but Android will then struggle for multiple frames trying to render it all with each update. The only theories or ideas I have:

- Maui is forcing excess redraws on updates
- Simple updates like translations that should be basically instantaneously and "free" are costing way too much time due to wrong functions being called.
- Simple things like Label text updates are doing way too much and taking too much time.

All of this is generally invisible until you have something scrolling because that is the only time every millisecond counts.

This is a good example of the type of bug that is likely doing this:

https://github.com/dotnet/maui/pull/21291

@jonathanpeppers found that the label text updates were taking way too long because they were getting bogged down in something very slow they didn't need to be doing.

What proof can we provide there are still other problems of this nature? I think we need to evaluate the performance of the core functions for basic inefficiencies like this because that is where the problem likely derives from. ie. Working from the bottom up. Ensure every component can do perform equivalently to Xamarin or native.

For example should we make .NET Android projects to compare against MAUI projects? I can make an infinite "autoscroller" in both that randomly updates the entries with labels, but this is only so useful as we don't have equivalent native Border. I don't know Xamarin so that would be a pain in the ass.

Certainly I think one needs to do torture tests of the basic elements against their native .NET or Xamarin equivalents before one can claim the issue is "CollectionView" and not primarily just bad performance in the basic elements. Eg. Have 500 Labels translating randomly on screen spamming with text updates also on a Timer set to 500 fps and see what you get on both for resulting deltaTime on the Timer as the system struggles.

**Comment by AliKarimiENT** [comment]
Besides this issue, we have faced a new one on release mode the GroupHeader binding mode doesn't work and we are surprised when we were getting ready to publish it.

**Comment by Phenek** [comment]
Furthermore, I would like to add that when we change the source of an embedded icon on each items, the scrolling performance on iOS decreases drastically.

As most layouts nowadays require customized images for each item, it would be wise to optimize this aspect.

**Comment by Equabyte** [comment]
CV rendering performance does affect scrolling in a rather noticeable way, there's simply too much going on on the main thread in a non-optimized way; in connection with an API loader this is far from ideal because it freezes the scroll view when rendering the next bunch of items di display. This is a cross-platform issue causing poor user experience that should be somehow prioritized.

**Comment by PureWeen** [comment]
We've made a number of fixes in SR6

Still keeping this open for SR7 to continue tracking improvements. 

If folks can test nightly on Android and report back that'd be helpful

Can you test with the latest nightly build? 
https://github.com/dotnet/maui/wiki/Nightly-Builds

**Comment by Equabyte** [comment]
> Can you test with the latest nightly build?
> https://github.com/dotnet/maui/wiki/Nightly-Builds
Hi @PureWeen I just tested on Android with 8.0.60-ci.net8.24301.1 and in my case I can see no noticeable improvement. Let me also clarify the actual issue I'm facing with scrolling performance refers to the case of a CollectionView that requires items to be added incrementally as the user scrolls (think of news items read from some API, preloaded and buffered in a non-visible data structure and incrementally moved to a CollectionView). The issue is that, whatever solution I tried to adopt so far, adding (or changing) a few items to the CollectionView causes a short UI freeze that stops the scrolling until the rendering is complete. There seems to be no way to get the CollectionView to update in background while the scrolling happens, so even when I try to entertain the user with empty placeholder items, the scrolling is affected by the UI update in a rather annoying manner.
Loading the entire CollectionView prior to showing it to user is not an option in these cases, because the initial wait would take too long, so an incremental loading mechanism is recommended.

**Comment by Domik234** [comment]
<details>
<summary>Basic Info</summary>
MAUI Version: 8.0.40
Configuration: Debug
Important to set to .csproj: `<UseInterpreter>false<UseInterpreter>`
</details>

<details>
<summary>Layout info</summary>

**CollectionView Test**

Grouping enabled.

DataTemplateSelector for HeaderTemplate, SubheaderTemplate, ItemTemplate.

**HeaderTemplate** - Label (bigger - NunitoSansBold)
**SubheaderTemplate** - Label (smaller - NunitoSansRegular)

**ItemTemplate** - Grid
Grid contains:
    - LeftView - TintedImage (Image with custom TintColor implementation)
    - Primary Text - One line only - NunitoSansRegular
    - Secondary Text - Two lines (filled with data from three string properties)
    -- First line 2 properties
    -- Second line 1 property
    -- NunitoSansRegular
    - RightView - Switching between TintedImage and Text

**Layout - 3 columns, 2 rows - rowspan = 2 for LeftView and RightView**
![image](https://github.com/dotnet/maui/assets/9479585/ac55a971-5c5e-4fbe-a0ed-cea76a8803d2)

**Examples**
![image](https://github.com/dotnet/maui/assets/9479585/f843a1d4-506c-43f3-8ac0-241b8ba34727)

</details>

<details>
<summary>Video (max 60 fps)</summary>

[CollectionViewTest.webm](https://github.com/dotnet/maui/assets/9479585/56b9e255-124c-4c77-be6b-25f6f20a8d44)

</details>

Tested on Honor 20. This feels fluent (in earlier versions there were problems primarily on iOS). Propably greater problem on much complicated layouts or weak devices.

**Comment by kimhongka** [comment]
Im also having this issue with infinit loading more logic. Also, I need to load items on top of items as Im building a chat box. I have 5different templates like image, video and text templates. Im migrating Xamarin.forms app to MAUI and Xamarin version used Listview for it , worked fine, but listview with same template and logic in MAUI was soooo laggy and also the item appearing checking logic in Listview has a bug that it dose not return the visible first item correctly at scrolling with Listview in MAUI, so I have replaced it to Collectionview, it works reasonably well on Android, but the same codes in iOS is quite slow, the inital loading taking 1-2seconds and scrolling is really jumppy. The logic itself works as Android works, but it is really laggy in iOS.... Please provide us a solution on this, really want to see MAUI is a successor version over Xamarin. Thank you.

**Comment by Equabyte** [comment]
One extra observation: starting from version 8.0.60 (or at least I don't recall having seen this in version 8.0.40) there is one additional bug in Android that affects CollectionView scrolling; not sure if this is the case in iOS because I could not test there so far. Basically, when a CV is nested in a ScrollView, after scrolling down real fast the loaded CV items (by swiping up with a fast gesture), the scrolling will correctly stop once the last CV item is fully displayed on screen. But now there's a bug that somehow prevents the "Scrolled" event of the ScrollView to fire again when the user attempts to scroll further down, to e.g. dynamically load more CV items which are not yet visible. This is not correct and we should let the "Scrolled" event to fire also in that situation, i.e. regardless of being already at the end of the CollectionView. It is up to the developer to then have appropriate logic in the "Scrolled" event method to handle the situation. Not sure if this bug is more on the CollectionView itself or the ScrollView though.

**Comment by xyeie** [comment]
When can we expect the performance issues of this collection view to be resolved? Please let us know. If it takes too long, we will need to switch to another cross-platform solution as soon as possible.

**Comment by xyeie** [comment]
When can we expect the performance issues of this collection view to be resolved? Please let us know. If it takes too long, we will need to switch to another cross-platform solution as soon as possible.

**Comment by davidortinau** [comment]
@xyeie @kimhongka @Equabyte @Phenek @AliKarimiENT please open an issue including a reproduction sample with your specific CollectionView related scenario. There's no one-size-fits-all issue or solution here, and we need focused issues to efficiently troubleshoot and provide solutions. Link this issue from the new issue if you feel they are related.

In reviewing the project on _this_ issue I find that all the samples perform well when published to device with the exception of a ColV with multiple bindable layouts (CardWithComplexContent) on Android. That sample could possibly be improved on iOS as well, but it scrolls very smoothly and we'll need to profile it to see if there's actually any stutter.

**Comment by Equabyte** [comment]
@davidortinau you are right in tackling one issue at a time. The sample @eth-ellis provided with the complex bindable content cards should be good enough to start investigation and it may shed some light to CV rendering performance on Android in general.

**Comment by Domik234** [comment]
Hello @KostGraz,

Have you tried disabling the interpreter using the `<UseInterpreter>false</UseInterpreter>` tag in .csproj? 
Have you tried to change ItemStrategy to "MeasureFirstItem"? (Looks like every item is the same size)

Is the layout of CollectionView items:
```
<Grid RowDefinitions="auto, auto" ColumnDefinitions="*">
<Label Grid.Row="0" Grid.Column="0" Text="...binding 1" />
<Label Grid.Row="1" Grid.Column="0" Text="...binding 2" />
</Grid>
```
? (I am worrying about usage of VerticalStackLayout)

Is this debug or release mode?

**Comment by Domik234** [comment]
@KostGraz thanks for the info!

The first step with UseInterpreter is the most important for speed if you don't want to use C# Hot Reload. For me it's like 5 times slower with this value set on true.

Other info I've asked is for potentional solving of MAUI team members. 😁

**Comment by jonmdev** [comment]
Just for reference, it seems @jsuarezruiz has recreated the same issue with another project here as of May 2024:

https://github.com/dotnet/maui/issues/22185

Is there any target in mind for when improving this might be a priority for the team? I think everyone would appreciate any help with it. This has been a continuously reported issue for a long time now.

Thanks for any help.

**Comment by CleverSoftwarePoland** [comment]
Any news or solid workaround? Problem with CollectionView also affects CarouselView any CarouselView with collection is sluggish even with simple templates.

**Comment by jonmdev** [comment]
> @xyeie @kimhongka @Equabyte @Phenek @AliKarimiENT please open an issue including a reproduction sample with your specific CollectionView related scenario. There's no one-size-fits-all issue or solution here, and we need focused issues to efficiently troubleshoot and provide solutions. Link this issue from the new issue if you feel they are related.
> 
> In reviewing the project on _this_ issue I find that all the samples perform well when published to device with the exception of a ColV with multiple bindable layouts (CardWithComplexContent) on Android. That sample could possibly be improved on iOS as well, but it scrolls very smoothly and we'll need to profile it to see if there's actually any stutter.

On what objective basis are you drawing this conclusion? OP's samples were created on Release mode and clearly show the deterioration in scroll function that we are all complaining about. It is visible in iOS and Android.

Are you suggesting this is not existing in the documented evidence there? You cannot see it in the videos or do not consider these videos important?

What standard is "perform well when published to device" when there is a clear and objective deterioration in function? Does this mean "when running on my high end test device faster than most users have and not under real world full app stress"?

I don't mean to sound rude, but I homestly feel like the primarily approach of the entire Maui team has been to gaslight us all about the poor layout performance issues both in iOS and Android for the past year.

No matter how many times anyone posts threads about it they are met with some combo of:

- actually Maui is really faster check out these amazing start up times.
- are you sure you're testing in release mode?
- I don't see the problem on my high tend top of the line test devices.
- please spend a week creating and posting another reproduction project (even though we have numerous). 

The problem clearly exists both in Android and iOS. Endless users are complaining about it. You have absolute objective evidence in the videos of the OP both for Android and iOS. @jsuarezruiz posted yet another Android project showing the issue clear as day with massive abnormal resource utilization in Maui layout.

I hope there is sufficient willpower to look at and fix the problem. I don't think there is insufficient evidence.

My theory remains that Maui is doing far too many layout calls on each update resulting in excess measurements and invalidating. This could be tested by overriding a platform view/handler and debugging out how many times on measure or something like that is called per update in Xamarin vs Maui.

Either way every bit of performance counts. Not every user has a high end test device. We should not be accepting dramatic objective deterioration in rendering function in Maui from Xamarin.

If the layout system in Maui is running in circles, these circles should be found and fixed. Thanks again for any help on this. It is becoming one of the biggest remaining Maui issues.

I hope you can agree that just repeating to us the same few bullet points above over and over for a year is not making Maui layouts faster.

**Comment by Alex-Dobrynin** [comment]
Buy the way guys. I've already released v1.1.7 of my VirtualizeListView, you can find it here https://github.com/MPowerKit/VirtualizeListView
I know it is not a silver bullet, but in some case it can save nerves and time.

**Comment by maonaoda** [comment]
> Buy the way guys. I've already released v1.1.7 of my VirtualizeListView, you can find it here https://github.com/MPowerKit/VirtualizeListView
> I know it is not a silver bullet, but in some case it can save nerves and time.

Will it be able to change the correct size based on the binding of header,footer,item (this is still a serious bug in maui for ios, but has not even been fixed. :(

**Comment by bcaceiro** [comment]
I have to agree with @jonmdev .

On a positive note, there have been significat improvements over the last past months regarding dotnet maui development and state.

However, the performance for rendering elements, with or without CollectionView, can still be improved. The proof of that is the work that @jonathanpeppers has been doing, with performance improvements all over different aspects. Even yesterday I was looking about 2 more, that impact the performance of everyone's application. Wish there were more people focused on this, or perhaps, in the current or next release , be a bigger priority. there are samples everywhere. Or, instead, get low end devices and see what's going on.

**Comment by bcaceiro** [comment]
Since it is hard to share the code of our applications most of the times, would it be preferrable to have a centralized  place, to submit all speedscopes to be analyzed in the future?

**Comment by Alex-Dobrynin** [comment]
@maonaoda yes, items can change the size based on binding context, also items can change size dynamically if the content size changes eg expander

**Comment by CleverSoftwarePoland** [comment]
I'm really struggling with performance problems on MAUI during our app migration from Xamarin 
I did create simple project representing mix of CarouselView and CollectionView showing performance problem. 
It's dead simple with two labels.
https://github.com/dotnet/maui/issues/23715

**Comment by jonmdev** [comment]
I did some preliminary testing on @eth-ellis 's project and found what I think were some very bizarre behaviors in the Maui version. I will keep you in suspense however, until I can test the Xamarin one to properly compare.

I tried opening Xamarin in Visual Studio 2022 and running it on Android but got:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error		file failed to compile.	CollectionViewPerformanceXamarin.Android	\\?\C:\Users\User\CollectionViewXamarinMaui\CollectionViewPerformanceXamarin\CollectionViewPerformanceXamarin\CollectionViewPerformanceXamarin.Android\obj\Debug\130\lp\60\jl\res\xml\xamarin_essentials_fileprovider_file_paths.xml		
Error		file failed to compile.	CollectionViewPerformanceXamarin.Android	C:\Users\User\CollectionViewXamarinMaui\CollectionViewPerformanceXamarin\CollectionViewPerformanceXamarin\CollectionViewPerformanceXamarin.Android\obj\Debug\130\lp\61\jl\res\layout\fallbacktoolbardonotuse.xml		
Error		file failed to compile.	CollectionViewPerformanceXamarin.Android	C:\Users\User\CollectionViewXamarinMaui\CollectionViewPerformanceXamarin\CollectionViewPerformanceXamarin\CollectionViewPerformanceXamarin.Android\obj\Debug\130\lp\61\jl\res\layout\fallbacktabbardonotuse.xml		
Error		layout_fallbacktabbardonotuse.xml.flat: error: failed to open file. "layout_fallbacktabbardonotuse.xml.flat: error: failed to open file.".	CollectionViewPerformanceXamarin.Android			
Error		layout_fallbacktoolbardonotuse.xml.flat: error: failed to open file. "layout_fallbacktoolbardonotuse.xml.flat: error: failed to open file.".	CollectionViewPerformanceXamarin.Android			
Error		xml_xamarin_essentials_fileprovider_file_paths.xml.flat: error: failed to open file. "xml_xamarin_essentials_fileprovider_file_paths.xml.flat: error: failed to open file.".	CollectionViewPerformanceXamarin.Android			
```

Does anyone know what that means? Can anyone else try downloading the OP repo and see if you can run it? Is there any obvious fix so I can run it? Thanks for any help.

I tried deleting all the obj and bin folders, rebuilding, cleaning, and I already have Xamarin installed:

![xamarin](https://github.com/user-attachments/assets/b17921d3-a29f-40f6-874c-3bc2c03a5b72)

How do I test the Xamarin project to compare?

**Comment by jonmdev** [comment]
I have forked the project from @eth-ellis and split it with my own edits to:

- https://github.com/jonmdev/CollectionViewPerformanceXamarin
- https://github.com/jonmdev/CollectionViewPerformanceMaui

If you are working with the Xamarin project, be advised that you must keep it in a root directory (eg. C:\CollectionViewPerformanceXamarin) as Xamarin otherwise starts giving weird errors if the paths get long. See https://learn.microsoft.com/en-us/answers/questions/313806/why-am-i-getting-this-error-on-a-new-xamarin-forms

I have only just started but I will continue this approach as I have time.

I put my custom code in folders called "MY CUSTOM CODE" so it is obvious what I did. I just created so far an override for the Android Label class in Xamarin and Maui so I could debug out when an OnMeasure or LayoutUpdate occurred.

I'm sure there will be many more problems. However, this one was just quite crazy and hit me right in the face.

## Problem: Maui project starts with an infinite measurement/layout feedback loop

Remember what I said about thinking the Layout system is the problem? 😂 When you start the Xamarin project this is what you get on debug:

```

[0:] MEASURE UPDATE 1 ID: 214177538
[0:] LAYOUT UPDATE 1
[0:] MEASURE UPDATE 2 ID: 214177538
[0:] LAYOUT UPDATE 2
[0:] MEASURE UPDATE 3 ID: 116333858
[0:] LAYOUT UPDATE 3
[0:] MEASURE UPDATE 4 ID: 47251612
[0:] LAYOUT UPDATE 4
[0:] MEASURE UPDATE 5 ID: 20679304
[0:] LAYOUT UPDATE 5
[0:] MEASURE UPDATE 6 ID: 100471506
[0:] LAYOUT UPDATE 6
[0:] MEASURE UPDATE 7 ID: 91589150
[0:] LAYOUT UPDATE 7
[0:] MEASURE UPDATE 8 ID: 57420330
[0:] LAYOUT UPDATE 8
[0:] MEASURE UPDATE 9 ID: 41598710
[0:] LAYOUT UPDATE 9
[0:] MEASURE UPDATE 10 ID: 112461799
[0:] LAYOUT UPDATE 10
[0:] MEASURE UPDATE 11 ID: 244608894
[0:] LAYOUT UPDATE 11
[0:] MEASURE UPDATE 12 ID: 42206602
[0:] LAYOUT UPDATE 12
[0:] MEASURE UPDATE 13 ID: 50488772
[0:] LAYOUT UPDATE 13
[0:] MEASURE UPDATE 14 ID: 77732400
[0:] LAYOUT UPDATE 14
[0:] MEASURE UPDATE 15 ID: 222734684
[0:] LAYOUT UPDATE 15
[0:] MEASURE UPDATE 16 ID: 105054024
[0:] LAYOUT UPDATE 16
```
This means there were 16 label measurements and 16 label layout updates. (I am just counting up for each measurement/layout that occurs). The "ID" is the HashId for the label just for reference. Totally normal. It stops there and stays idle.

By contrast if you start the Maui project it goes into an endless loop of measuring and laying out such that after a few seconds, the Debug output is:

```

[0:] MEASURE UPDATE 2753 ID: 130619213
[0:] MEASURE UPDATE 2754 ID: 99036496
[0:] MEASURE UPDATE 2755 ID: 18898778
[0:] MEASURE UPDATE 2756 ID: 132098689
[0:] MEASURE UPDATE 2757 ID: 173764884
[0:] MEASURE UPDATE 2758 ID: 125422339
[0:] MEASURE UPDATE 2759 ID: 118667178
[0:] LAYOUT UPDATE 1221
[0:] MEASURE UPDATE 2760 ID: 130619213
[0:] LAYOUT UPDATE 1222
[0:] MEASURE UPDATE 2761 ID: 99036496
[0:] LAYOUT UPDATE 1223
[0:] MEASURE UPDATE 2762 ID: 18898778
[0:] LAYOUT UPDATE 1224
[0:] MEASURE UPDATE 2763 ID: 132098689
[0:] LAYOUT UPDATE 1225
[0:] MEASURE UPDATE 2764 ID: 173764884
[0:] LAYOUT UPDATE 1226
[0:] MEASURE UPDATE 2765 ID: 125422339
[0:] LAYOUT UPDATE 1227
[0:] MEASURE UPDATE 2766 ID: 116333858
[0:] MEASURE UPDATE 2767 ID: 116333858
[0:] LAYOUT UPDATE 1228
[0:] MEASURE UPDATE 2768 ID: 38820601
[0:] MEASURE UPDATE 2769 ID: 137909740
[0:] MEASURE UPDATE 2770 ID: 15590166
[0:] MEASURE UPDATE 2771 ID: 62381677
[0:] MEASURE UPDATE 2772 ID: 177567472
[0:] MEASURE UPDATE 2773 ID: 154763151
[0:] MEASURE UPDATE 2774 ID: 38820601
[0:] MEASURE UPDATE 2775 ID: 137909740
[0:] MEASURE UPDATE 2776 ID: 15590166
[0:] MEASURE UPDATE 2777 ID: 62381677
[0:] MEASURE UPDATE 2778 ID: 177567472
[0:] MEASURE UPDATE 2779 ID: 154763151
[0:] MEASURE UPDATE 2780 ID: 19362727
[0:] LAYOUT UPDATE 1229
[0:] MEASURE UPDATE 2781 ID: 38820601
[0:] LAYOUT UPDATE 1230
[0:] MEASURE UPDATE 2782 ID: 137909740
[0:] LAYOUT UPDATE 1231
[0:] MEASURE UPDATE 2783 ID: 15590166
[0:] LAYOUT UPDATE 1232
[0:] MEASURE UPDATE 2784 ID: 62381677
[0:] LAYOUT UPDATE 1233
[0:] MEASURE UPDATE 2785 ID: 177567472
[0:] LAYOUT UPDATE 1234
[0:] MEASURE UPDATE 2786 ID: 154763151
[0:] LAYOUT UPDATE 1235
[0:] MEASURE UPDATE 2787 ID: 130619213
[0:] MEASURE UPDATE 2788 ID: 99036496
[0:] MEASURE UPDATE 2789 ID: 18898778
[0:] MEASURE UPDATE 2790 ID: 132098689
[0:] MEASURE UPDATE 2791 ID: 173764884
[0:] MEASURE UPDATE 2792 ID: 125422339
[0:] MEASURE UPDATE 2793 ID: 130619213
[0:] MEASURE UPDATE 2794 ID: 99036496
[0:] MEASURE UPDATE 2795 ID: 18898778
[0:] MEASURE UPDATE 2796 ID: 132098689
[0:] MEASURE UPDATE 2797 ID: 173764884
```

It is endlessly spamming measurements and layouts. Slightly more than 2:1 measurements to layout updates. It will go up to the tens of thousands if you leave it. You can also notice the label hash ids are all over the place indicating likely it is not just updating but killing and creating labels constantly in this death spiral.

You can break the feedback loop by scrolling down far enough. But this is absolutely insane on startup.

I don't believe this is the main problem, but it is certainly a problem, and some evidence for my general theory that the layout system is running in circles too often.

**Comment by jonmdev** [comment]
I further just updated the projects to add some similar basic iOS label update debugging. iOS fortunately does not seem to go into the death spiral update loop on start up in Maui. However, it demonstrates another problem.

## Problem 2 (iOS): Maui draws labels 41% more times than Xamarin for the same workload

Again, this goes back to what I have said has seemed blatantly obvious to me - Maui is overdoing simple work due to the layout system. Here we can objectively and reproduceably measure that **Maui draws labels in iOS 42% more than Xamarin for the same workflow.**

### Test Method
Open the project in Maui/Xamarin (whichever to start with). Let it load. Then click to select in the menu "CardWithComplexContent". Start scrolling down the screen at your own pace until you hit the bottom.

Look at the count of label draws incremented in the Debug output. I did this twice in Xamarin and twice in Maui and got identical results both times on my test device.

Maui Initial Load:
```
[0:] LAYOUT UPDATE 1
[0:] LAYOUT UPDATE 2
[0:] LAYOUT UPDATE 3
[0:] LAYOUT UPDATE 4
[0:] LAYOUT UPDATE 5
[0:] LAYOUT UPDATE 6
[0:] LAYOUT UPDATE 7
[0:] LAYOUT UPDATE 8
[0:] LAYOUT UPDATE 9
[0:] LAYOUT UPDATE 10
[0:] LAYOUT UPDATE 11
[0:] LAYOUT UPDATE 12
[0:] LAYOUT UPDATE 13
[0:] LAYOUT UPDATE 14
[0:] LAYOUT UPDATE 15
[0:] LAYOUT UPDATE 16
[0:] LAYOUT UPDATE 17
```

Maui Final Count:

TRIAL 1:
[0:] LAYOUT UPDATE 1321
[0:] LAYOUT UPDATE 1322
[0:] LAYOUT UPDATE 1323
[0:] LAYOUT UPDATE 1324

TRIAL 2:
[0:] LAYOUT UPDATE 1321
[0:] LAYOUT UPDATE 1322
[0:] LAYOUT UPDATE 1323
[0:] LAYOUT UPDATE 1324

Xamarin INitial Load
[0:] LAYOUT UPDATE 1
Thread started:  #11
[0:] LAYOUT UPDATE 2
[0:] LAYOUT UPDATE 3
[0:] LAYOUT UPDATE 4
[0:] LAYOUT UPDATE 5
[0:] LAYOUT UPDATE 6
[0:] LAYOUT UPDATE 7
[0:] LAYOUT UPDATE 8
[0:] LAYOUT UPDATE 9
[0:] LAYOUT UPDATE 10
[0:] LAYOUT UPDATE 11
[0:] LAYOUT UPDATE 12
[0:] LAYOUT UPDATE 13
[0:] LAYOUT UPDATE 14
[0:] LAYOUT UPDATE 15

Xamarin Final Count

TRIAL 1:
[0:] LAYOUT UPDATE 929
[0:] LAYOUT UPDATE 930
[0:] LAYOUT UPDATE 931
[0:] LAYOUT UPDATE 932
[0:] LAYOUT UPDATE 933

TRIAL 2:
[0:] LAYOUT UPDATE 929
[0:] LAYOUT UPDATE 930
[0:] LAYOUT UPDATE 931
[0:] LAYOUT UPDATE 932
[0:] LAYOUT UPDATE 933

### Conclusion

Thus we see consistently in both trials Xamarin needed 933 label updates to load, switch page, and scroll to the bottom, while in both trials Maui needed 1324 label updates. That is **42% more label updates to accomplish the same thing on the same device.**

Making sense now?

**Comment by jonmdev** [comment]
I updated the test projects once more to add a Reset Counters button so I could test Android's performance. 

This will not be quite as perfectly reproducable as we are not starting from identical conditions each time given the need to escape the initial feedback loop. However, the data is just as stark as you will see.

## Problem 3 (Android): Maui performs 62% more Label measurements than Xamarin to perform the same workload

### Test Method
Run the project in Android debug to device. Then click Change Template > CardWIthComplexContent.

Scroll down until the feedback loop stops (it stops spamming updates), then scroll back to the top. Click Reset Counters to go back to zero on the update counters. Clear your debug output also perhaps. Then scroll down through the full list with this as your "starting point".

### Results

**Maui:**

_TRIAL 1:_
```
[0:] MEASURE UPDATE 3006 
[0:] LAYOUT UPDATE 1224
```
_TRIAL 2:_
```
[0:] MEASURE UPDATE 2994 
[0:] LAYOUT UPDATE 1222
```

**Xamarin:**

_TRIAL 1:_
```
[0:] MEASURE UPDATE 1882
[0:] LAYOUT UPDATE 1264
```
_TRIAL 2:_
```
[0:] MEASURE UPDATE 1824 
[0:] LAYOUT UPDATE 1248
```

### Conclusion

Thus we see an average of 1853 Label Measure commands in Xamarin vs. 3000 in Maui. 

**This represents 62% more Label measurement commands in Maui vs. Xamarin to accomplish the same thing.**

Layout calls were reasonably similar by contrast.

Is this stark enough? Is it still any mystery why we are suffering for performance both in Android and iOS?

**Comment by jonmdev** [comment]
I have always felt the Image management in Maui has been quite poor. Loading simple images in a "CollectionView" type situation (ie. while scrolling) always leads to stutters and seems far too arduous no matter how much you cache or create platform Bitmaps first.

So I extended the project to add an Image View. I just copied and pasted the "ComplexCard" of the project into "CardWithPhoto" and threw some random cat photos in there.

The outcomes are absolutely disastrous and shocking. I don't usually use XAML (I code in C#) so unless there is something I missed here, while Xamarin handled the change fine for both Android and iOS, it's an absolute wreck in Maui across the board.

### Test Method
I added a random photo into the XAML like so with some code elsewhere to select and add one of 14 cat photos at that spot:

```
            <StackLayout
                Margin="10"
                Spacing="10">

                <Label
                    Text="{Binding Source.RestaurantName, Source={x:Reference this}}"
                    FontSize="Title" />

                <Label
                    Text="{Binding Source.RestaurantDescription, Source={x:Reference this}}"
                    FontSize="Body" />

                <Image 
                    Source="{Binding Source.PhotoImageSource, Source= {x:Reference this}}" 
                    />

                <Label
                    Text="{Binding Source.RestaurantAddress, Source={x:Reference this}}"
                    TextColor="DodgerBlue"
                    FontAttributes="Bold"
                    FontSize="Body" />

```

It looked quite nice and laid out just right in Xamarin iOS and Android:

![xamarin Android](https://github.com/user-attachments/assets/17a194d7-8310-4e6d-9983-dbf19273ef70) ![xamarin iOS](https://github.com/user-attachments/assets/abfed257-870c-46b0-b406-f27f4e3a2536)

However this is what we got on Maui iOS and Android. Android was an abomination where there was endless space above and below the images. iOS just truncated the bottom of the views. Both were grotesque by comparison to Xamarin.

![maui Android 1](https://github.com/user-attachments/assets/ca84eaf9-ff18-436f-9113-3cf6362315d0)
![maui Android 2](https://github.com/user-attachments/assets/23737554-a78a-4769-81b1-8309f702a40c)

![maui iOS 1](https://github.com/user-attachments/assets/ab064428-8ad9-4ad1-bd6e-e0c955d8edf0)
![maui iOS 2](https://github.com/user-attachments/assets/26b77836-b256-42bd-91a8-46874016f453)

I am not sure again on XAML and how much might have changed to Maui, but to me there is no reason this should fail so differently on iOS and Android. Clearly something is broken.

So let's call this:

## Problem 4 (Android & iOS): Neither can accurately lay out the same XAML templates as Xamarin when Images are involved

Though this is cosmetic. The abomination goes deeper and it gets weirder.

**Comment by jonmdev** [comment]
## Problem 4 (Android): Android in Maui calls `OnDraw` at least 21x for each Image State Update

The Android Image data set was difficult to test in Maui because Maui has some terrifying problems. First of all, it was very hard to get out of the startup infinite update loop once the images were added into the mix. Furthermore, scrolling back upwards frequently led to errors as I will point out in a minute.

However, one of the most strikingly strange things I encountered in all this is the following.

With Android images, we have:

```
        protected override void DrawableStateChanged() {
            UpdateCounter.addImageDrawStateUpdate(); 
            base.DrawableStateChanged();
        }
        protected override void OnDraw(Canvas canvas) {
            UpdateCounter.addImageDrawUpdate(); //doesn't run on xamarin at all
            base.OnDraw(canvas);
        }

        protected override void OnMeasure(int widthMeasureSpec, int heightMeasureSpec) {
            UpdateCounter.addImageMeasureUpdate(this.GetHashCode().ToString());
            base.OnMeasure(widthMeasureSpec, heightMeasureSpec);
        }
        protected override void OnLayout(bool changed, int left, int top, int right, int bottom) {
            UpdateCounter.addImageLayoutUpdate();
            base.OnLayout(changed, left, top, right, bottom);
        }
```

What was fascinating and bizarre is the Xamarin manages to render the photos without once calling `OnDraw`. It only calls `DrawableStateChanged` and roughly in proportion to how many photos it must display.

**Xamarin:**

Here is the output from Xamarin Android on (1) Loading app, (2) Clear Counters, (3) Select CardWithPhoto, (4) Scroll to bottom:

_TRIAL 1:_
```
[0:] IMAGE STATE UPDATE 119
[0:] IMAGE MEASURE UPDATE 177
[0:] IMAGE LAYOUT UPDATE 177
```

_TRIAL 2:_
```
[0:] IMAGE STATE UPDATE 121
[0:] IMAGE MEASURE UPDATE 177
[0:] IMAGE LAYOUT UPDATE 177
```

For reference there should be 100 images, so it is only running image state changed slightly more than the total count of images. It doesn't run `OnDraw` even once.

**Maui:**

By contrast, in Maui, it is an absolute disaster. I was not able to scroll through the page due to the infinite loop and crashes, but I could gather a bit of data in that by one point I had before crashing:

```
[0:] IMAGE STATE UPDATE 21
[0:] IMAGE DRAW UPDATE 455
```

This suggests that by the 21st photo display or so, it had run `OnDraw` 455 times (drawn 455 photos?), which is absolutely surreal. How is Xamarin not running `OnDraw` at all? While Maui is running it 21x more often than there are image state changes???

It blows my mind.

**Comment by jonmdev** [comment]
## Problem 5 (Android): Android in Maui crashes due to its "unique" drawing method compared to Xamarin

It seems clear then that Image is now drawing photos in a completely different way from Xamarin and this is not favorable at all. The component seems deeply broken compared to the Xamarin version.

As you scroll you will encounter endless errors of:
`[View] ContentViewGroup not displayed because it is too large to fit into a software layer (or drawing cache), needs 12932556 bytes, only 10368000 available`

(note that Xamarin handles the exact same photos without a single complaint.)

And eventually if you scroll back up you will crash with:

```
[Bitmap] Called getConfig() on a recycle()'d bitmap! This is undefined behavior!
**Java.Lang.RuntimeException:** 'Canvas: trying to use a recycled bitmap android.graphics.Bitmap@236f184'
```

Where to begin? 😬

Any thoughts?

**Comment by jonmdev** [comment]
## Problem 7 (iOS): Maui lays out Images ~2.7 times as many times as Xamarin for the same workload

With iOS the image behavior is at least not so broken as to be unquantifiable, but just like everything else it is far poorer than Xamarin.

Loading the app, clearing the counters, Loading the CardWithPhoto, and then scrolling top to bottom I get:

**Maui:**

_TRIAL 1:_

```
[0:] IMAGE LAYOUT UPDATE 406
```

_TRIAL 2:_

```
[0:] IMAGE LAYOUT UPDATE 396
```

**Xamarin:**

_TRIAL 1:_

```
[0:] IMAGE LAYOUT UPDATE 138
```

_TRIAL 2:_
```
[0:] IMAGE LAYOUT UPDATE 154
```
Thus we get an average of 146 image layouts in Xamarin, and 401 image layouts in Maui, meaning **Maui lays out Images 2.7x as many times as Xamarin.**

What more can be said?

😬

**Comment by jonmdev** [comment]
Final Thoughts?

**Comment by bcaceiro** [comment]
Great work @jonmdev , seems reasonably obvious that there is an issue, regarding overdrawing that will affect all applications. Hope this gets the right attention asap

**Comment by Alex-Dobrynin** [comment]
@jonmdev You can just look at their commits, to understand if they need mobile frmaework:
tests, tests, tests, ui tests, windows windows windows, snapshots, tests tests, windows...
And NO, Absolutely nothing related to layout problems on ios and android.

**Comment by AlleSchonWeg** [comment]
This should be number one prio (top of tops). The layout system was not the fastest in XF but Maui is a lot slower.
Perhaps the team should concentrate to mobile and analyse why layouting, measuring and drawing cycles are increasing.

The reactions on @jonmdev 's tests clearly showing that a lot of people observe similar behaviors. I know this is difficult but we (and our customers) need a performant app.

**Comment by jonmdev** [comment]
## Update: Simple Layout Test

I was curious and wanted to see if (1) the measuring/layout/drawing issues can be reproduced by a very simple layout configuration and (2) if problems would be amplified by increasing depth of nesting layout objects.

I edited my projects here:

- https://github.com/jonmdev/CollectionViewPerformanceXamarin
- https://github.com/jonmdev/CollectionViewPerformanceMaui

By changing App.xaml.cs to allow an alternative application build. This is switched by the bool which remains defaulted to the original CollectionView project. If set true, you will run the simple new project instead.

This project just adds a single or nest of AbsoluteLayout (or StackLayout or VerticalStackLayout) elements and then a Label at the end in the deepest of the hierarchy. Then it clears the update counter on startup and issues one resize command to the hierarchy of layouts.

A perfectly efficient system should only give one label layout and draw after the resize.

### RESULTS

**Android:**
- Xamarin - 1 measure, 1 layout (perfect efficiency) 
- Maui - 3 measures, 2 layout (2-3x less efficient than perfect). 
- Same result with both AbsoluteLayout and StackLayout/VerticalStackLayout.

**iOS:**
- With AbsoluteLayout - 2 layouts, 1 draw (same imperfection in Xamarin & Maui). 
- With StackLayout/VerticalStackLayout - 3 layouts and 1 draw in Maui and 2 layouts and 1 draw in Xamarin (less efficient in Maui).

**Nesting:**
- There is no difference whether you have just one layout or nest up to 10+. It does not amplify the problem.

### CONCLUSIONS
Layout problems in Maui can be very easily reproduced without any "Bindings" or CollectionView both in Android and iOS. It only takes around 10-30 lines of C# code to reproduce.

Maui inefficiencies in Android are seen equally with both AbsoluteLayout and StackLayout/VerticalStackLayout in this simple test.

Maui inefficiencies in iOS are also reproduced with VerticalStackLayout/StackLayout. I believe there are AbsoluteLayout inefficiencies and problems in iOS as well (as I use them regularly) though I can't reproduce it in this simple test. I will need to think and test more about this.

### IDEA TO SOLVE
If I was you guys (Microsoft staff), I would take a simple project like this one, add some Debug outs inside the Maui measuring/layout system at the various steps and see which steps are being run too many times in each OS and why. 

I would guess this would be the easiest way to figure out the problems. Using simple C# like here you can ramp up the complexity of the layouts as things get solved and compare. Then go to the CollectionView project for eventual testing of real world performance. 

Perfect CollectionView performance can be measured by counting the number of labels/images in each entry and multiplying by the number of elements (100 by default). So if 8 labels per element, once counter is cleared, then CollectionView is loaded from menu, then scrolled to bottom, perfect performance would be 800 label layouts/measures/draws.

This is all quantifiable in such simple terms. Perhaps there are other ways to approach this. That is just the best I can think of.

### TEST CODE

App.xaml.cs is updated in each to:

```
public App() {

    //===================
    //toggle projects
    //===================
    bool testSimpleLayout = false;

    //============================================
    //1) ORIGINAL PROJECT USING MAUI SYSTEM
    //============================================
    if (!testSimpleLayout) {
        InitializeComponent();
        MainPage = new AppShell();
    }
    //============================================
    //2) SIMPLE PROJECT USING ABS LAYOUT
    //============================================
    else {
        ContentPage mainPage = new();
        MainPage = mainPage;

        int numNested = 10; //set number of extra layouts to nest from 0-n, does not change result

        List<VerticalStackLayout> layoutList = new(); //can try alternative layout types here
        VerticalStackLayout layout = new(); //can try alternative layout types here

        mainPage.Content = layout;
        mainPage.Content = layout;
        layoutList.Add(layout);
        
        for (int i=0; i < numNested; i++) {
            layout = new();
            layoutList[layoutList.Count - 1].Add(layout);
            layoutList.Add(layout);
        }

        Label label = new();
        label.Text = "HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO HELLO";
        layoutList[layoutList.Count-1].Add(label);

        mainPage.SizeChanged += delegate {
            UpdateCounter.resetCounters();
            Debug.WriteLine("====================RESET COUNTERS ON RESIZE");
            if (mainPage.Width > 0) {
                for (int i = 0; i < layoutList.Count; i++) {
                    layoutList[i].WidthRequest = mainPage.Width;
                    layoutList[i].HeightRequest = mainPage.Height;
                }
            }
        };

        //ABSOLUTE LAYOUT:
        //== android maui - measure 3x, layout 2x | android xamarin - measure 1x, layout 1x
        //== iOS maui - layout 2x, draw 1x | iOS xamarin - layout 2x, draw 1x

        //VERTICAL STACK LAYOUT & STACK LAYOUT:
        //== android maui - measure 3x, layout 2x | android xamarin - measure 1x, layout 1x
        //== iOS maui - layout 3x, draw 1x | iOS xamarin - layout 2x, draw 1x
    }

}
```

Thanks for your ongoing efforts.

**Comment by Equabyte** [comment]
Improving CollectionView performance is the one single most convincing MAUI enhancement that will produce the highest short-term reward as literally anybody creating professional apps with complex layouts will benefit from it: I see this card keeps switching across product increments but it's worth considering for prioritization.

**Comment by chrisg32** [comment]
I am glad this one has been assigned but I would love to see the `p/1` tag on it.

With all the [open verified performance issues for android](https://github.com/dotnet/maui/issues?q=is%3Aopen+label%3A%22platform%2Fandroid+%F0%9F%A4%96%22+label%3At%2Fperf+label%3As%2Fverified+label%3At%2Fbug) we are very hesitant on releasing our app to production. We were on a time crunch to migrate from Xamarin Forms to MAUI support Android 15. Since Android 15 is expected to be released today, there will be customers we will be unable to support.

**Comment by jonmdev** [comment]
Looks like @brentpbc posted another good example of the problem for iOS here:

https://github.com/dotnet/maui/issues/24224

The original title for this current SR9 issue was "[iOS/Android] CollectionView scrolling performance worse in .NET MAUI when compared to Xamarin.Forms (lag/stutter)"

I notice the current issue title for this thread was edited to remove iOS by @PureWeen in June seemingly by @davidortinau 's suggestion.

I don't mean to rub it in or again be rude or abrasive, but as @PureWeen is the current assignee and @davidortinau is the Microsoft head of Maui, is it currently recognized by the Maui team that an issue exists with both Android and iOS? Will both be fixed?

I hope performance in both will be corrected. No one is using Maui with the intention of only developing mobile for Android. In my own application I have worse performance in iOS actually at this point.

@davidortinau you also asked for more examples. Looks like another very good one there in that linked report for iOS showing the same issue.

**Comment by bcaceiro** [comment]
No news on this subject?

**Comment by simon-biber** [comment]
.NET 9 Preview 7 has [CollectionView & CarouselView improvements with a new opt-in handler for iOS and Mac Catalyst](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#collectionview--carouselview-improvements-with-a-new-opt-in-handler-for-ios-and-mac-catalyst)

> This release introduces two new handlers for developers to try that bring sweeping performance and stability improvements for both CollectionView and CarouselView. These new implementations are based on newer UICollectionView APIs. Opt-in by adding the following into your Program.cs:
```
#if IOS || MACCATALYST
appBuilder.ConfigureMauiHandlers(handlers =>
{
  handlers.AddHandler<Microsoft.Maui.Controls.CollectionView, Microsoft.Maui.Controls.Handlers.Items2.CollectionViewHandler2>();
  handlers.AddHandler<Microsoft.Maui.Controls.CarouselView, Microsoft.Maui.Controls.Handlers.Items2.CarouselViewHandler2>();
});
#endif
```

**Comment by chrisg32** [comment]
> .NET 9 Preview 7 has [CollectionView & CarouselView improvements with a new opt-in handler for iOS and Mac Catalyst](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#collectionview--carouselview-improvements-with-a-new-opt-in-handler-for-ios-and-mac-catalyst)
> 
> > This release introduces two new handlers for developers to try that bring sweeping performance and stability improvements for both CollectionView and CarouselView. These new implementations are based on newer UICollectionView APIs. Opt-in by adding the following into your Program.cs:
> 
> ```
> #if IOS || MACCATALYST
> appBuilder.ConfigureMauiHandlers(handlers =>
> {
>   handlers.AddHandler<Microsoft.Maui.Controls.CollectionView, Microsoft.Maui.Controls.Handlers.Items2.CollectionViewHandler2>();
>   handlers.AddHandler<Microsoft.Maui.Controls.CarouselView, Microsoft.Maui.Controls.Handlers.Items2.CarouselViewHandler2>();
> });
> #endif
> ```

That is well in good for iOS but those of us needing an immediate stable release for Android are still stuck.

**Comment by jonmdev** [comment]
> .NET 9 Preview 7 has [CollectionView & CarouselView improvements with a new opt-in handler for iOS and Mac Catalyst](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#collectionview--carouselview-improvements-with-a-new-opt-in-handler-for-ios-and-mac-catalyst)
> 
> > This release introduces two new handlers for developers to try that bring sweeping performance and stability improvements for both CollectionView and CarouselView. These new implementations are based on newer UICollectionView APIs. Opt-in by adding the following into your Program.cs:
> 
> ```
> #if IOS || MACCATALYST
> appBuilder.ConfigureMauiHandlers(handlers =>
> {
>   handlers.AddHandler<Microsoft.Maui.Controls.CollectionView, Microsoft.Maui.Controls.Handlers.Items2.CollectionViewHandler2>();
>   handlers.AddHandler<Microsoft.Maui.Controls.CarouselView, Microsoft.Maui.Controls.Handlers.Items2.CarouselViewHandler2>();
> });
> #endif
> ```

Thanks. That's interesting but not of any particular benefit to me. I do not use CollectionView but rather my own custom systems. Again, the problem is not CollectionView. The problem is Maui's layout system.

Xamarin was fast using the old API. As proven by numerous projects and reports. Which again shows that the problem is in the layout system of Maui, not the API.

There are numerous projects and tests showing that Maui's layout system is defective. I spent almost a week straight doing tests and experiments to try to illustrate it here.

Can you acknowledge whether you guys recognize this issue as such and are working on trying to fix the basic layout issues in both iOS and Android and attain the same performance as Xamarin had?

I think we all understand this is hard work but we would all very much appreciate it. It is demoralizing to be working in a system like Maui where we can be micromanaging our own code optimizations but in the background Maui is still sluggish because the fundamental layouts are doubling and tripling the OS's workload.

Thanks again for your help and interest in the problem.

**Comment by AlleSchonWeg** [comment]
Hi @jonmdev
perhaps you should create a new issue with your tests. Hopefully the new issue will get more attention than the comments in this issue.

**Comment by lucapan** [comment]
@Alex-Dobrynin thanks a lot !!!

**Comment by jonmdev** [comment]
For cross referencing, I believe I have found the primary cause of the poor scrolling/collectionview/translation performance in iOS Maui. 

Every time you change the translation of any objects in iOS, Maui is causing the whole hierarchy to re-arrange itself and re-measure itself in iOS. This is not similarly happening in Android/Windows.

This matches my experience where the poorest performing systems I now have are in iOS. It makes sense as any custom scrolling or translation I perform is spamming massively expensive re-measurements and re-arrangements which permeate down through the hierarchy.

I have shared my bug report here which demonstrates the issue: https://github.com/dotnet/maui/issues/24996 

As far as a solution, @PureWeen, any thoughts?

**Comment by taublast** [comment]
guess nb1: Could benchmark applying default styles in Xamarin vs MAUI
guess nb2: Could benchmark measuring text for a specifically set font in Xamarin vs MAUI
guess nb3: there have been many CollectionView fixes, at some point it might have introduced an additional triggered re-measurement non-existing in Xamarin.

**Comment by PureWeen** [comment]
The original scenarios that were logged here all appear to be resolved with the latest releases of .NET MAUI

I've tested through a few other scenarios and those also appear to be resolved.

If you're still having performance issues with .NET MAUI SR9 (which did include some pretty decent performance fixes) please log a new issue so we can track and resolve
